cpp/src/copying/contiguous_split.cu

```diff
@@ -813,282 +907,298 @@ struct size_of_helper {
-      out.element_size  = in.element_size;
-      out.value_shift   = in.value_shift;
-      out.bit_shift     = in.bit_shift;
-      out.valid_count =
-        in.valid_count;  // valid count will be set to 1 if this is a validity buffer
-      out.src_buf_index = in.src_buf_index;
-      out.dst_buf_index = in.dst_buf_index;
+  size_type* h_indices;
+  src_buf_info* h_src_buf_info;
 
-      size_type const elements_per_chunk =
-        out.element_size == 0 ? 0 : chunk_size / out.element_size;
-      out.num_elements = ((chunk_index + 1) * elements_per_chunk) > in.num_elements
-                           ? in.num_elements - (chunk_index * elements_per_chunk)
-                           : elements_per_chunk;
+  int offset_stack_partition_size;
+  size_type* d_indices;
+  src_buf_info* d_src_buf_info;
+  size_type* d_offset_stack;
+};
 
-      size_type const rows_per_chunk =
-        // if this is a validity buffer, each element is a bitmask_type, which
-        // corresponds to 32 rows.
-        out.valid_count > 0
-          ? elements_per_chunk * static_cast<size_type>(detail::size_in_bits<bitmask_type>())
-          : elements_per_chunk;
-      out.num_rows = ((chunk_index + 1) * rows_per_chunk) > in.num_rows
-                       ? in.num_rows - (chunk_index * rows_per_chunk)
-                       : rows_per_chunk;
+// packed block of memory 2: partition buffer sizes and dst_buf_info structs
+struct packed_partition_buf_size_and_dst_buf_info {
+  packed_partition_buf_size_and_dst_buf_info(cudf::table_view const& input,
+                                             std::vector<size_type> const& splits,
+                                             std::size_t num_partitions,
+                                             cudf::size_type num_src_bufs,
+                                             std::size_t num_bufs,
+                                             rmm::cuda_stream_view stream,
+                                             rmm::mr::device_memory_resource* mr)
+    : stream(stream),
+      buf_sizes_size{cudf::util::round_up_safe(num_partitions * sizeof(std::size_t), split_align)},
+      dst_buf_info_size{cudf::util::round_up_safe(num_bufs * sizeof(dst_buf_info), split_align)}
+  {
+    // host-side
+    h_buf_sizes_and_dst_info = std::vector<uint8_t>(buf_sizes_size + dst_buf_info_size);
+    h_buf_sizes              = reinterpret_cast<std::size_t*>(h_buf_sizes_and_dst_info.data());
+    h_dst_buf_info =
+      reinterpret_cast<dst_buf_info*>(h_buf_sizes_and_dst_info.data() + buf_sizes_size);
+
+    // device-side
+    d_buf_sizes_and_dst_info = rmm::device_buffer(buf_sizes_size + dst_buf_info_size, stream, mr);
+    d_buf_sizes              = reinterpret_cast<std::size_t*>(d_buf_sizes_and_dst_info.data());
+
+    //// destination buffer info
+    d_dst_buf_info = reinterpret_cast<dst_buf_info*>(
+      static_cast<uint8_t*>(d_buf_sizes_and_dst_info.data()) + buf_sizes_size);
+  }
```

Comment:
I think these can be added to the ctor initialization?
```suggestion
    // host-side
    h_buf_sizes_and_dst_info{std::vector<uint8_t>(buf_sizes_size + dst_buf_info_size)},
    h_buf_sizes{reinterpret_cast<std::size_t*>(h_buf_sizes_and_dst_info.data())},
    h_dst_buf_info{reinterpret_cast<dst_buf_info*>(h_buf_sizes_and_dst_info.data() + buf_sizes_size)},
    // device-side
    d_buf_sizes_and_dst_info{rmm::device_buffer(buf_sizes_size + dst_buf_info_size, stream, mr)},
    d_buf_sizes{reinterpret_cast<std::size_t*>(d_buf_sizes_and_dst_info.data())},
    // destination buffer info
    d_dst_buf_info{reinterpret_cast<dst_buf_info*>(static_cast<uint8_t*>(d_buf_sizes_and_dst_info.data())+buf_sizes_size)}
```
This may allow some of the member variables to be `const` perhaps.

Reply:
If `h_buf_sizes_and_dst_info` and `d_buf_sizes_and_dst_info` are non-`const`, the rest can be made `const` pretty cleanly.  

The reason is the `reinterpret_cast`. Since `.data()` would be `T * const` I would cast away the constness. I could `reinterpret_cast<K>(const_cast<K>(x.data()))` to allow me to make the contiguous vector/device buffer const, but then I need to "Bring back the const" (:tm:) with another `const_cast`.

I'll go the route of all members in the initializer, and as many `const` as the compiler lets me and it is fairly clean, unless I hear otherwise, but open for other thoughts here.

Reply:
Sounds good to me.

Reply:
I tried `h_buf_sizes_and_dst_info{std::vector<uint8_t>(buf_sizes_size + dst_buf_info_size)}`, but I am getting a narrowing compiler error (`-Werror=narrowing`) that I don't fully understand:

```
cudf/cpp/src/copying/contiguous_split.cu: In constructor 'cudf::_GLOBAL__N__eeb54488_19_contiguous_split_cu_64d0a9ad_16515::packed_split_indices_and_src_buf_info::packed_split_indices_and_src_buf_info(const cudf::table_view&, const std::vector<int>&, std::size_t, cudf::size_type, rmm::cuda_stream_view, rmm::mr::device_memory_resource*)':
cudf/cpp/src/copying/contiguous_split.cu:959:242: error: narrowing conversion of '(((long unsigned int)((int)((cudf::_GLOBAL__N__eeb54488_19_contiguous_split_cu_64d0a9ad_16515::packed_split_indices_and_src_buf_info*)this)->cudf::_GLOBAL__N__eeb54488_19_contiguous_split_cu_64d0a9ad_16515::packed_split_indices_and_src_buf_info::indices_size)) + ((long unsigned int)((cudf::_GLOBAL__N__eeb54488_19_contiguous_split_cu_64d0a9ad_16515::packed_split_indices_and_src_buf_info*)this)->cudf::_GLOBAL__N__eeb54488_19_contiguous_split_cu_64d0a9ad_16515::packed_split_indices_and_src_buf_info::src_buf_info_size))' from 'long unsigned int' to 'unsigned char'
```

 I can't seem to repro it myself, so I am not entirely sure why it is so confused here.

This, however, works:

```
h_indices_and_source_info(indices_size + src_buf_info_size)
```

Reply:
Patch here: https://github.com/rapidsai/cudf/pull/13260/commits/0e10fdc901594154b47da2f87e453b1095f1427c